engine/src/test/java/io/zeebe/engine/processor/workflow/multiinstance/MultiInstanceActivityTest.java

```diff
@@ -539,28 +539,67 @@ public void shouldCollectNestedOutputElements() {
             .withVariable(INPUT_COLLECTION_EXPRESSION, INPUT_COLLECTION)
             .create();
 
-    RecordingExporter.jobRecords(JobIntent.CREATED)
-        .withWorkflowInstanceKey(workflowInstanceKey)
-        .limit(INPUT_COLLECTION.size())
-        .exists();
-
-    // complete job
+    // complete jobs
     completeJobs(
         workflowInstanceKey,
         INPUT_COLLECTION.size(),
         i -> Map.of("nested", OUTPUT_COLLECTION.get(i)));
 
     // then
-    final List<Long> scopeKeys =
-        RecordingExporter.variableRecords(VariableIntent.CREATED)
-            .withName(OUTPUT_ELEMENT_EXPRESSION)
-            .map(Record::getValue)
-            .map(VariableRecordValue::getScopeKey)
-            .limit(3)
-            .collect(Collectors.toList());
-    assertThat(scopeKeys)
-        .hasSize(3)
-        .allSatisfy(scopeKey -> assertThat(scopeKeys).containsOnlyOnce(scopeKey));
+    assertThat(
+            RecordingExporter.variableRecords(VariableIntent.CREATED)
+                .withName(OUTPUT_ELEMENT_EXPRESSION) // without '.nested'
+                .withValue("null")
+                .limit(INPUT_COLLECTION.size()))
+        .hasSize(INPUT_COLLECTION.size());
+
+    assertThat(
+            RecordingExporter.variableRecords()
+                .withName(OUTPUT_COLLECTION_VARIABLE)
+                .withScopeKey(workflowInstanceKey)
+                .getFirst()
+                .getValue())
+        .hasValue(JsonUtil.toJson(OUTPUT_COLLECTION));
+  }
+
+  @Test
+  public void shouldCollectOutputElementsFromExpression() {
+    // given
+    ENGINE
+        .deployment()
+        .withXmlResource(
+            workflow(
+                miBuilder.andThen(
+                    m ->
+                        m.zeebeOutputElementExpression(
+                            "number(string(loopCounter) + string(loopCounter))"))))
+        .deploy();
+
+    final long workflowInstanceKey =
+        ENGINE
+            .workflowInstance()
+            .ofBpmnProcessId(PROCESS_ID)
+            .withVariable(INPUT_COLLECTION_EXPRESSION, INPUT_COLLECTION)
+            .create();
+
+    // complete jobs
+    completeJobs(workflowInstanceKey, INPUT_COLLECTION.size());
+
+    // then
+    assertThat(
+            RecordingExporter.variableRecords(VariableIntent.CREATED)
+                .map(Record::getValue)
+                .map(VariableRecordValue::getName)
+                .limit(10))
```

Comment:
Limiting the records to a specific number is a bit unsteady. Instead, we can limit the records from creating a workflow instance to its completion/termination.

```suggestion
             RecordingExporter.records()
                .limitToWorkflowInstance(workflowInstanceKey)
                .variableRecords()
                .map(Record::getValue)
                .map(VariableRecordValue::getName))
```

Reply:
Nice, that's a good one to learn about.